06.wc/wc.rb

```diff
@@ -0,0 +1,100 @@
+#!/usr/bin/env ruby
+
+# frozen_string_literal: true
+
+require 'optparse'
+
+def main
+  option = ARGV.getopts('l')
+  file_paths = ARGV
+
+  if file_paths.empty?
+    count_word(option)
+  else
+    count_word_of_argv(option, file_paths)
+  end
+end
+
+def count_word(option)
+  text = readlines.join
+  file = nil
+  word_count_info = []
+  word_count_info << count_file_info(file, text)
+
+  padded_result = create_padded_result(option, word_count_info)
+
+  puts padded_result.join
+end
+
+def count_word_of_argv(option, file_paths)
+  count_info_list = build_count_info_list(file_paths)
+  count_info_list << build_total_hash(count_info_list) if file_paths.size > 1
+  padded_result = create_padded_result(option, count_info_list)
+  padded_result.each { |result_array| puts result_array.join(' ') }
+end
+
+def build_total_hash(count_info_list)
+  total_hash = {}
+  total_hash[:rows_count] = 0
+  total_hash[:words_count] = 0
+  total_hash[:bytes] = 0
+
+  count_info_list.map do |data_in_hash|
+    total_hash[:rows_count] += data_in_hash[:rows_count]
+    total_hash[:words_count] += data_in_hash[:words_count]
+    total_hash[:bytes] += data_in_hash[:bytes]
+  end
+  total_hash[:file_name] = 'total'
+
+  total_hash
+end
+
+def build_count_info_list(file_paths)
+  ret = []
+  file_paths.each do |file_path|
+    if File.exist?(file_path)
+      if FileTest.directory?(file_path)
+        puts "wc: #{file_path}: read: Is a directory"
+      else
+        text = File.read(file_path)
+        ret << count_file_info(file_path, text)
+      end
+    else
+      puts "wc: #{file_path}: open: No such file or directory"
+    end
+  end
+  ret
```

Comment:
上級テクニックですが、「`puts`は`nil`を返す＝偽」で「`count_file_info(file_path, text)`はハッシュを返す＝真」なので、真の要素だけを配列の要素にして返してくれる`filter_map`メソッドが使えそうです（今のコードのままでも全然OKです）。

```suggestion
  file_paths.filter_map do |file_path|
    if File.exist?(file_path)
      if FileTest.directory?(file_path)
        puts "wc: #{file_path}: read: Is a directory"
      else
        text = File.read(file_path)
        count_file_info(file_path, text)
      end
    else
      puts "wc: #{file_path}: open: No such file or directory"
    end
  end
```

https://docs.ruby-lang.org/ja/latest/method/Enumerable/i/filter_map.html